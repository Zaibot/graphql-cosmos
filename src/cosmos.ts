import { SqlParameter } from '@azure/cosmos';
import { GraphQLCosmosRequest } from './context';
import { SqlBuilder } from './sql';

export const defaultOnQuery = ({ client, database, container, query, parameters }: GraphQLCosmosRequest) =>
    client.database(database).container(container).items.query({ query, parameters }).fetchAll();

export function createSqlQuery(
    whereExpressions: Array<{ property: string; operation: string; value: unknown; parameter: string }>,
    sortExpressions: Array<{ property: string; direction: string }>,
) {
    const fromAlias = `c`;
    const expressions = whereExpressions.map((expr) => {
        const operationMap: Record<string, string> = {
            '': `${fromAlias}.${expr.property} = ${expr.parameter}`,
            eq: `${fromAlias}.${expr.property} = ${expr.parameter}`,
            neq: `${fromAlias}.${expr.property} != ${expr.parameter}`,
            gt: `${fromAlias}.${expr.property} > ${expr.parameter}`,
            gte: `${fromAlias}.${expr.property} >= ${expr.parameter}`,
            lt: `${fromAlias}.${expr.property} < ${expr.parameter}`,
            lte: `${fromAlias}.${expr.property} <= ${expr.parameter}`,
            in: `ARRAY_CONTAINS(${expr.parameter}, ${fromAlias}.${expr.property})`,
            nin: `NOT ARRAY_CONTAINS(${expr.parameter}, ${fromAlias}.${expr.property})`,
            contains: `CONTAINS(${fromAlias}.${expr.property}, ${expr.parameter})`,
            ncontains: `NOT CONTAINS(${fromAlias}.${expr.property}, ${expr.parameter})`,
        };
        const sql: string = operationMap[expr.operation];
        const parameter: SqlParameter = { name: expr.parameter, value: expr.value as any };
        if (sql) {
            return { sql, parameter };
        } else {
            throw Error(`unknown expression in ${JSON.stringify(whereExpressions)}`);
        }
    });

    const sql = new SqlBuilder(fromAlias);
    for (const expr of expressions) {
        if (expr) {
            sql.where(expr.sql);
        }
    }

    for (const { property, direction } of sortExpressions) {
        if (direction === `ASC`) {
            sql.orderBy(`${fromAlias}.${property}`, `ASC`);
        } else if (direction === `DESC`) {
            sql.orderBy(`${fromAlias}.${property}`, `DESC`);
        } else {
            throw Error(`sort direction of ${property} must be ASC or DESC`);
        }
    }

    sql.orderBy(`${fromAlias}.id`, `ASC`);

    const parameters = expressions.map((x) => x?.parameter).filter((x): x is SqlParameter => !!x);

    return { sql, parameters };
}
