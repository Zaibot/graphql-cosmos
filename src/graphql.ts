import { GraphQLInputFieldConfig, GraphQLSchema, GraphQLInputType, GraphQLInputObjectType, GraphQLField } from 'graphql';

export const createOrGetWhereType = (name: string, properties: Record<string, GraphQLInputFieldConfig>, schema: GraphQLSchema) => {
    if (schema.getTypeMap()[name]) {
        return schema.getTypeMap()[name] as GraphQLInputType;
    }

    const filterType = new GraphQLInputObjectType({
        name,
        description: undefined,
        extensions: undefined,
        fields: properties,
        astNode: undefined,
        extensionASTNodes: undefined,
    });

    // HACK
    schema.getTypeMap()[name] = filterType;

    return filterType;
};

export const addFieldArgument = (field: GraphQLField<any, any, any>, name: string, type: GraphQLInputType) => {
    field.args.push({
        name,
        type,
        astNode: undefined,
        extensions: undefined,
        description: undefined,
        defaultValue: undefined,
    });
};
