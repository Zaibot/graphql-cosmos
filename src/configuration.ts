import { CosmosClient, FeedOptions, FeedResponse } from '@azure/cosmos'
import { DataLoaderHandler } from './dataloader/loader'
import { CosmosRequest } from './intermediate/model'
import { SqlBuilder } from './sql/builder'

export interface GraphQLCosmosContext {
  directives: GraphQLDirectivesContext
}

export type CosmosQueryHandler = (request: GraphQLCosmosRequest) => Promise<FeedResponse<any>>

export interface GraphQLDirectivesContext {
  cosmos: {
    database: string
    client: CosmosClient
    dataloader?: DataLoaderHandler<GraphQLCosmosContext>
    /** default: defaultOnInit */
    onInit?: (request: CosmosRequest, init: GraphQLCosmosInitRequest) => void
    onBeforeQuery?: (request: GraphQLCosmosInitRequest) => void
    /** default: defaultOnQuery */
    onQuery?: CosmosQueryHandler
  }
}

export interface GraphQLCosmosInitRequest {
  client: CosmosClient
  request: CosmosRequest
  database: string
  container: string
  query?: SqlBuilder
  parameters?: Array<{ name: string; value: any }>
  options?: FeedOptions
}

export interface GraphQLCosmosRequest {
  init?: GraphQLCosmosInitRequest
  client: CosmosClient
  database: string
  container: string
  query: string
  parameters: Array<{ name: string; value: any }>
  options?: FeedOptions
}
