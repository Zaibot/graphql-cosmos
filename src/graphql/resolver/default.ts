import * as GraphQL from 'graphql'
import { SchemaVisitorMap, visitSchema } from 'graphql-tools'
import { debugHooks } from '../../debug'
import { cosmosOursByResolveInfo, cosmosTheirsByResolveInfo } from '../directive/ast'
import { hasCosmosTag } from '../reference'
import { hasId, requireCosmosColumn } from './requireCosmosColumn'

export function buildCosmosFieldResolverSchema(schema: GraphQL.GraphQLSchema) {
  const map: SchemaVisitorMap = {
    'VisitSchemaKind.OBJECT_TYPE'(type) {
      for (const field of Object.values(type.getFields())) {
        debugHooks?.onResolverSet?.({
          resolver: `requireCosmosColumn`,
          objectType: type,
          fieldType: field,
        })

        field.resolve = DefaultCosmosFieldResolver(field)
      }
      return type
    },
  }
  const schema2 = visitSchema(schema, map)
  return schema2
}

function DefaultCosmosFieldResolver(field: GraphQL.GraphQLField<any, any, { [key: string]: any }>) {
  const nextResolver = field.resolve ?? GraphQL.defaultFieldResolver
  const cosmosResolver: typeof nextResolver = async (s: unknown, a, c, i) => {
    if (hasCosmosTag(s) && hasId(s) && !cosmosTheirsByResolveInfo(i)) {
      const ours = cosmosOursByResolveInfo(i)
      const column = ours ?? field.name
      const sourced = await requireCosmosColumn(s, column, c)
      const result = await nextResolver(sourced, a, c, i)
      return result
    } else {
      return await nextResolver(s, a, c, i)
    }
  }
  return cosmosResolver
}
