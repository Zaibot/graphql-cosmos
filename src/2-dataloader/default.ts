import { createDataLoader, DataLoaderHandler } from './loader'
import { defaultOnDataLoaderResolve } from './resolver'

export const defaultDataLoader = (): DataLoaderHandler => {
  const loader = createDataLoader({})
  return loader
}
