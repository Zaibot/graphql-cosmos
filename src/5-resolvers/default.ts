import { GraphQLCosmosFieldResolver } from './resolver'
import { SourceDescriptor } from './x-descriptors'

export const defaultCosmosFieldResolver: GraphQLCosmosFieldResolver = async (parent, _args, context, info) => {
  const field = context.dataSources.graphqlCosmos.meta.requireField(info.parentType.name, info.fieldName)
  const column = field.ours ?? field.fieldname
  const source = SourceDescriptor.getDescriptor(parent)
  if (source?.kind === `Single`) {
    const value = await context.dataSources.graphqlCosmos.load(source, column)
    return value
  } else {
    const value = Object(parent)[column]
    return value
  }
}

// export const defaultCosmosWhereFieldResolver: GraphQLCosmosFieldResolver = (parent, _args, context, info) => {
//   const field = context.dataSources.graphqlCosmos.meta.requireField(info.parentType.name, info.fieldName)
//   const column = field.whereOurs ?? field.ours ?? field.fieldname
//   const source = SourceDescriptor.getDescriptor(parent)
//   if (source?.kind === `Single`) {
//     return context.dataSources.graphqlCosmos.load(source, column)
//   } else {
//     const value = Object(parent)[column]
//     console.log(`EMBEDDED ${info.parentType}.${info.fieldName} = ${value}`)
//     return value
//   }
// }

// export const defaultCosmosSortFieldResolver: GraphQLCosmosFieldResolver = (parent, _args, context, info) => {
//   const field = context.dataSources.graphqlCosmos.meta.requireField(info.parentType.name, info.fieldName)
//   const column = field.sortOurs ?? field.ours ?? field.fieldname
//   const source = SourceDescriptor.getDescriptor(parent)
//   if (source?.kind === `Single`) {
//     return context.dataSources.graphqlCosmos.load(source, column)
//   } else {
//     const value = Object(parent)[column]
//     return value
//   }
// }
