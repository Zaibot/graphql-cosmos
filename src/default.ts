import { GraphQLCosmosInitRequest, GraphQLCosmosRequest } from './configuration'
import { createDataLoader, DataLoaderHandler } from './dataloader/loader'
import { defaultOnDataLoaderResolve } from './dataloader/resolver'
import { CosmosRequest } from './intermediate/model'
import { convertToSql } from './sql'

export const defaultOnQuery = ({ client, database, container, query, parameters, options }: GraphQLCosmosRequest) =>
  client.database(database).container(container).items.query({ query, parameters }, options).fetchNext()

export const defaultOnInit = (request: CosmosRequest, init: GraphQLCosmosInitRequest) => {
  const sql = convertToSql(request)
  init.query = sql.sql
  init.parameters = sql.parameters
  init.options ??= {}
  init.options.continuationToken = request.cursor
}

export const defaultDataLoader = (): DataLoaderHandler => {
  const loader = createDataLoader({ resolve: defaultOnDataLoaderResolve })
  return loader
}
