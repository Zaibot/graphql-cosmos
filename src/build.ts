import * as GraphQL from 'graphql'
import { SchemaDirectiveVisitor } from 'graphql-tools'
import { GraphQLCosmosSchema } from './graphql/directive/schema'
import { buildCosmosFieldResolverSchema as applyCosmosDefaultFieldResolvers } from './graphql/resolver/default'

export const buildCosmosSchema = (
  schema: GraphQL.GraphQLSchema,
  directives: Record<string, typeof SchemaDirectiveVisitor> = GraphQLCosmosSchema.schemaDirectives
): GraphQL.GraphQLSchema => {
  const extended = new GraphQL.GraphQLSchema(schema.toConfig())
  SchemaDirectiveVisitor.visitSchemaDirectives(extended, directives)
  applyCosmosDefaultFieldResolvers(schema)

  const cleaned = extended.toConfig()
  cleaned.directives = cleaned.directives.filter((t) => !Object.keys(directives).includes(t.name))

  const finished = new GraphQL.GraphQLSchema(cleaned)
  return finished
}

export const buildCosmosASTSchema = (
  typeDefs: GraphQL.DocumentNode,
  directives: Record<string, typeof SchemaDirectiveVisitor> = GraphQLCosmosSchema.schemaDirectives
): GraphQL.GraphQLSchema => {
  return buildCosmosSchema(
    GraphQL.buildASTSchema(GraphQL.concatAST([typeDefs, GraphQLCosmosSchema.typeDefs])),
    directives
  )
}
