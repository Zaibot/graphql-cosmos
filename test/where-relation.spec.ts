import { FeedResponse } from '@azure/cosmos'
import { execute, GraphQLSchema, printSchema, validate, validateSchema } from 'graphql'
import gql from 'graphql-tag'
import { buildCosmosASTSchema } from '../src/build'
import { GraphQLCosmosContext, GraphQLCosmosRequest } from '../src/configuration'
import { defaultDataLoader } from '../src/default'

const dummyTypeDefs = gql`
  type Query {
    entities: [Entity!]! @cosmos(container: "Entities")
  }

  type Entity {
    id: ID!
    related: Related @cosmos(container: "Related", ours: "relatedId") @where(op: "eq", ours: "relatedId")
  }

  type Related {
    id: ID!
    text: String
  }
`

describe(`Where`, () => {
  let context: GraphQLCosmosContext
  let dummy: GraphQLSchema

  beforeEach(() => {
    const loader = defaultDataLoader()

    context = {
      directives: {
        cosmos: {
          database: null as any,
          client: null as any,
          onQuery: onCosmosQuery,
          dataloader: loader,
        },
      },
    }

    dummy = buildCosmosASTSchema(dummyTypeDefs)
  })

  it(`schema`, async () => {
    const output = printSchema(dummy, { commentDescriptions: false })
    expect(validateSchema(dummy)).toHaveLength(0)
    expect(output).toMatchSnapshot()
  })

  it(`should retrieve all`, async () => {
    const query = gql`
      query {
        entities {
          page {
            __typename
            id
            related {
              __typename
              id
              text
            }
          }
        }
      }
    `
    expect(validate(dummy, query)).toHaveLength(0)

    const result = await execute(dummy, query, undefined, context)
    result.errors?.forEach((e) => console.error(e.path.join(`/`), e.stack))

    expect(result).toEqual({
      data: {
        entities: {
          page: [
            {
              __typename: 'Entity',
              id: `1`,
              related: {
                __typename: 'Related',
                id: `1a`,
                text: `a`,
              },
            },
            {
              __typename: 'Entity',
              id: `2`,
              related: {
                __typename: 'Related',
                id: `2b`,
                text: `b`,
              },
            },
          ],
        },
      },
    })
  })

  it(`should retrieve 1a`, async () => {
    const query = gql`
      query {
        entities(where: { relatedId_eq: "1a" }) {
          page {
            __typename
            id
            related {
              __typename
              id
              text
            }
          }
        }
      }
    `
    expect(validate(dummy, query)).toHaveLength(0)

    const result = await execute(dummy, query, undefined, context)
    result.errors?.forEach((e) => console.error(e.path.join(`/`), e.stack))

    expect(result).toEqual({
      data: {
        entities: {
          page: [
            {
              __typename: 'Entity',
              id: `1`,
              related: {
                __typename: 'Related',
                id: `1a`,
                text: `a`,
              },
            },
          ],
        },
      },
    })
  })
})

async function onCosmosQuery(request: GraphQLCosmosRequest): Promise<FeedResponse<unknown>> {
  const { container, query, parameters } = request
  const key = parameters.length ? `${query} (${parameters.map((x) => `${x.name}=${x.value}`).toString()})` : query

  const responses: Record<string, Record<string, unknown[]>> = {
    Entities: {
      'SELECT c.id FROM c ORDER BY c.id': [{ id: `1` }, { id: `2` }],
      'SELECT c.id, c.relatedId FROM c WHERE ARRAY_CONTAINS(@batch, c.id) (@batch=1,2)': [
        { __typename: 'Entity', id: `1`, relatedId: `1a` },
        { __typename: 'Entity', id: `2`, relatedId: `2b` },
      ],
      'SELECT c.id FROM c WHERE c.relatedId = @relatedId_eq ORDER BY c.id (@relatedId_eq=1a)': [
        { __typename: 'Entity', id: `1` },
      ],
      'SELECT c.id, c.relatedId FROM c WHERE ARRAY_CONTAINS(@batch, c.id) (@batch=1)': [
        { __typename: 'Entity', id: `1`, relatedId: `1a` },
      ],
    },
    Related: {
      'SELECT c.id FROM c ORDER BY c.id': [{ id: `1a` }, { id: `2b` }],
      'SELECT c.id, c.text FROM c WHERE ARRAY_CONTAINS(@batch, c.id) (@batch=1a,2b)': [
        { __typename: 'Related', id: `1a`, text: `a` },
        { __typename: 'Related', id: `2b`, text: `b` },
      ],
      'SELECT c.id, c.text FROM c WHERE ARRAY_CONTAINS(@batch, c.id) (@batch=1a)': [
        { __typename: 'Related', id: `1a`, text: `a` },
      ],
    },
  }

  const result = responses[container]?.[key]
  if (result) {
    return { resources: result } as any
  } else {
    throw Error(`Unhandled: ${container} ${key}`)
  }
}
