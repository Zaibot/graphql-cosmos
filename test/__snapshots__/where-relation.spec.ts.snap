// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Where schema 1`] = `
"type Query {
  entities(where: EntityWhere, cursor: String): EntityPage!
}

type Entity {
  id: ID!
  related: Related
}

type Related {
  id: ID!
  text: String
}

input EntityWhere {
  relatedId_eq: ID
}

type EntityPage {
  nextCursor: String
  total: Int!
  page: [Entity!]!
}
"
`;
